### ISSUE TITLE
[core] NPE in core.items.GenericItem -> bad parametrized logging message?

### ISSUE THREAD

**Issue by cweitkamp** [eclipse/smarthome#4104]
log in openHAB 2.1.0 - release build -

```
2017-08-21 12:32:14.893 [WARN ] [pse.smarthome.core.items.GenericItem] - failed notifying listener '[org.eclipse.smarthome.core.persistence.internal.PersistenceManagerImpl@165b978, netatmoOutdoorTempTrend, null]' about state update of item java.lang.NullPointerException: {}
```

log in openHAB 2.1.0 -  snapshot 1020 -
```
2017-08-21 21:19:47.965 [WARN ] [pse.smarthome.core.items.GenericItem] - failed notifying listener '[org.eclipse.smarthome.core.persistence.internal.PersistenceManagerImpl@2a31d596, Netatmo_EV_MasterBedroom_Luchtvochtigheid, null]' about state update of item {}: {}
java.lang.NullPointerException: null
	at org.mapdb.DataOutput2.writeUTF(DataOutput2.java:147) [198:org.openhab.persistence.mapdb:1.11.0.201708190109]
	at org.openhab.persistence.mapdb.internal.MapDBitemSerializer.serialize(MapDBitemSerializer.java:45) [198:org.openhab.persistence.mapdb:1.11.0.201708190109]
	at org.openhab.persistence.mapdb.internal.MapDBitemSerializer.serialize(MapDBitemSerializer.java:1) [198:org.openhab.persistence.mapdb:1.11.0.201708190109]
	at org.mapdb.BTreeMap$NodeSerializer.serialize(BTreeMap.java:397) [198:org.openhab.persistence.mapdb:1.11.0.201708190109]
	at org.mapdb.BTreeMap$NodeSerializer.serialize(BTreeMap.java:288) [198:org.openhab.persistence.mapdb:1.11.0.201708190109]
	at org.mapdb.Store.serialize(Store.java:154) [198:org.openhab.persistence.mapdb:1.11.0.201708190109]
	at org.mapdb.StoreWAL.update(StoreWAL.java:371) [198:org.openhab.persistence.mapdb:1.11.0.201708190109]
	at org.mapdb.Caches$HashTable.update(Caches.java:270) [198:org.openhab.persistence.mapdb:1.11.0.201708190109]
	at org.mapdb.EngineWrapper.update(EngineWrapper.java:63) [198:org.openhab.persistence.mapdb:1.11.0.201708190109]
	at org.mapdb.BTreeMap.put2(BTreeMap.java:707) [198:org.openhab.persistence.mapdb:1.11.0.201708190109]
	at org.mapdb.BTreeMap.put(BTreeMap.java:643) [198:org.openhab.persistence.mapdb:1.11.0.201708190109]
	at org.openhab.persistence.mapdb.internal.MapDBPersistenceService.store(MapDBPersistenceService.java:161) [198:org.openhab.persistence.mapdb:1.11.0.201708190109]
	at org.openhab.core.persistence.internal.PersistenceServiceDelegate.store(PersistenceServiceDelegate.java:55) [195:org.openhab.core.compat1x:2.2.0.201708201100]
	at org.eclipse.smarthome.core.persistence.internal.PersistenceManagerImpl.handleStateEvent(PersistenceManagerImpl.java:119) [103:org.eclipse.smarthome.core.persistence:0.9.0.201708151402]
	at org.eclipse.smarthome.core.persistence.internal.PersistenceManagerImpl.stateChanged(PersistenceManagerImpl.java:420) [103:org.eclipse.smarthome.core.persistence:0.9.0.201708151402]
	at org.eclipse.smarthome.core.items.GenericItem$1.run(GenericItem.java:229) [98:org.eclipse.smarthome.core:0.9.0.201708151402]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

Related discussion in community: https://community.openhab.org/t/what-channel-types-are-supported-by-wich-netatmo-device/24394/28

**Comment by kaikreuzer** [comment]
Hm, the exception occurs in `org.openhab.persistence.mapdb` - shouldn't it hence be reported at https://github.com/openhab/openhab1-addons or what makes you think that it is in any way related to ESH?

**Comment by cweitkamp** [comment]
I was not sure if this error is a concatenation of multiple circumstances or a specific situation for the mapdb persistence.

First of all the parametrized logging message has a bad format - one, maybe two, parameters is/are missing:
https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GenericItem.java#L232

After that I looked into the binding itself and it seems like one is able to assign `null` values to e.g. `StringTypes`. After a short investigation - looking into some unit tests - it shows up that this is possible and maybe expected. Isn't it?
So we could fix it in the binding itself.

Finally the persistence was the next step on my todo list. But I am not very experienced with it.

We of course can move the issue to https://github.com/openhab/openhab1-addons.

**Comment by kaikreuzer** [comment]
> First of all the parametrized logging message has a bad format 

Ok, true. This should be fixed with https://github.com/eclipse/smarthome/pull/4106.

> one is able to assign null values to e.g. StringTypes

You mean as a state to the item? No, this should not be allowed.
So imho the MapDBPersistenceService should be hardened to not crash in such a situation. And the binding should be fixed not to set null states.

**Comment by cweitkamp** [comment]
> Ok, true. This should be fixed with #4106.

Thanks.

> You mean as a state to the item? No, this should not be allowed.

I hope I understand it correctly. I mean something like this `new StringType(null)`:
https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/library/types/StringTypeTest.java#L41

> So imho the MapDBPersistenceService should be hardened to not crash in such a situation. And the binding should be fixed not to set null states.

I will take care of that.

**Comment by kaikreuzer** [comment]
> I hope I understand it correctly. I mean something like this new StringType(null)

Ah, that's indeed an interesting question. The tests indeed suggest that it is ok that way, but having passed null to it indeed means that this implementation crashes on many method calls.
Imho we should disallow null as a parameter or internally set it to empty in such a case. Interestingly, there are even tests that null should not be the same as empty - not sure if this really makes sense. @sjka Any opinion?

**Comment by sjka** [comment]
Reopening (was auto-closed by the PR)

**Comment by cweitkamp** [comment]
Very beasic setup to reproduce this error:

**Precondition:**
`mapdb` persistence installed and Item `mapdbString` should have a matching strategy : everyChange.

demo.items
```csv
Switch mapdbSwitch
String mapdbString "[%s]"
```

demo.rules
```php
rule "mapdb Test"
when
    Item mapdbSwitch changed
then
    mapdbString.sendCommand(new StringType(null))
end
```

**Comment by sjka** [comment]
Now thinking about it, IMHO the `StringType(null)` should be disallowed completely. It does not add any value - for that matter we have `UnDefType`.

It looks like I added these awkward assertions to the test with #2479. If I remember correctly, I mainly did it to verify that `StringType` _can_ handle null values gracefully. But in this PR I obviously didn't solve the root-cause, I was just hiding it.

For compatibility (and convenience) reasons I'd be okay though with a `StringType(...)` constructor which turns `null` into an empty string. Nevertheless, it remains true that we should try avoiding creating such null StringTypes in the first place, i.e. fix all the bindings that we can identify.

**Comment by maggu2810** [comment]
I would prefer to disallow the null type for the StringType constructor and not adding workarounds to the framework that handle wrong usage.
Sure we could state that to create an empty StringType you are allowed to call `StringType(null)` but does it make sense?

**Comment by kaikreuzer** [comment]
My main concern is that I cannot assess what this might break - as we see from this report, there null seems to be currently used and it was allowed for years... And it hasn't been a wrong usage, because it is not "null" as a state, but StringType instance, which is in general ok. So if we want to break the functionality, we should have a good reason.

> Sure we could state that to create an empty StringType you are allowed to call StringType(null) but does it make sense?

Well, it would be just another option to using `StringType.EMPTY` or `new StringType()` - and having `new StringType(null)` behaving the same way as `new StringType()` is imho quite ok and expected.

**Comment by triller-telekom** [comment]
> and having new StringType(null) behaving the same way as new StringType() is imho quite ok and expected

I would also add a `@NonNull` annotation in this constructor: `StringType(@NonNull String s) `. This will raise a warning everywhere where its called and people can then change their code to use `StringType()` instead.

**Comment by kaikreuzer** [comment]
@triller-telekom Please read https://github.com/eclipse/smarthome/issues/4104#issuecomment-323947528 above - it is used in scripts, so this does not only break code at compile time, but also individual user installations. So why should we force a NonNull, if there is no strong reason to do so?

**Comment by triller-telekom** [comment]
Sorry, I overlooked that its being used during runtime too.

**Comment by cweitkamp** [comment]
Just for the records: While testing #4963 I found another situation which should be handled in a different way:

https://github.com/eclipse/smarthome/blob/da57eda4edd9d41bbf41316f23e1b9b29687069b/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/DecimalType.java#L54-L56

The constructor [BigDecimal(String val)](https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html#BigDecimal-java.lang.String) throws a `NumberFormatException` if `val` is not a valid representation of a `BigDecimal`. I suggest to catch the exception and return `DecimalType.ZERO` or `new DecimalType()`.

**Comment by triller-telekom** [comment]
To sum this issue up:

- The logging error was fixed in #4106 which this issue was originally about
- As @kaikreuzer states in https://github.com/eclipse/smarthome/issues/4104#issuecomment-324031598 we probably should keep the `new StringType(null)` as a valid use case to not break any old stuff
- the `DecimalType` discussion brought up by @cweitkamp in https://github.com/eclipse/smarthome/issues/4104#issuecomment-361009347 is similar to this issue but not at all reflected by the original topic of this issue. @cweitkamp Do you want to create a new issue for that and close this one?

**Comment by cweitkamp** [comment]
FTR #5439
